core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala

```diff
@@ -0,0 +1,239 @@
+package org.ensime.core.javac
+
+import com.sun.source.tree.{ MemberSelectTree, MethodInvocationTree, Tree, IdentifierTree }
+import com.sun.source.util.TreePath
+import javax.lang.model.`type`.{ DeclaredType, PrimitiveType, ReferenceType, TypeKind, TypeMirror }
+import javax.lang.model.element.{ Element, ExecutableElement, PackageElement, TypeElement, VariableElement }
+import javax.lang.model.util.{ ElementFilter, Elements }
+import org.ensime.core.{ DocFqn, DocSig, DocSigPair, CompletionUtil }
+import org.ensime.util.FileUtils
+import org.ensime.api._
+import scala.collection.JavaConversions._
+import java.nio.charset.Charset
+import com.sun.source.tree.Scope
+import scala.collection.mutable.HashSet;
+import scala.collection.mutable.ArrayBuffer;
+
+trait JavaCompletion { self: JavaCompiler =>
+
+  import CompletionUtil._
+
+  def completionsAt(info: SourceFileInfo, offset: Int, maxResultsArg: Int, caseSens: Boolean): CompletionInfoList = {
+    val s = contentsAsString(info, charset)
+
+    val preceding = s.slice(Math.max(0, offset - 100), offset)
+
+    System.err.println("PRECEDING: " + preceding)
+
+    val defaultPrefix = JavaIdentRegexp.findFirstMatchIn(preceding) match {
+      case Some(m) => m.group(1)
+      case _ => ""
+    }
+
+    System.err.println("PREFIX: " + defaultPrefix)
+
+    val constructing = ConstructingRegexp.findFirstMatchIn(preceding).isDefined
+
+    System.err.println("CONSTRUCTING: " + constructing)
+
+    val indexAfterTarget = offset - defaultPrefix.length - 1
+
+    val precedingChar = s(indexAfterTarget)
+
+    val isMemberAccess = precedingChar == '.'
+
+    (if (ImportSubtypeRegexp.findFirstMatchIn(preceding).isDefined) {
+      // Erase the trailing partial subtype (it breaks type resolution).
+      val patched = s.substring(0, indexAfterTarget) + " " + s.substring(indexAfterTarget + defaultPrefix.length + 1);
+      (pathToPoint(SourceFileInfo(info.file, Some(patched), None), indexAfterTarget - 1) map {
+        case (info: CompilationInfo, path: TreePath) => {
+          memberCandidates(info, path.getLeaf, defaultPrefix, true, caseSens)
+        }
+      })
+    } else if (ImportRegexp.findFirstMatchIn(preceding).isDefined) {
+      (pathToPoint(info, indexAfterTarget) flatMap {
+        case (info: CompilationInfo, path: TreePath) => {
+          getEnclosingMemberSelectTree(path).map { m =>
+            packageMemberCandidates(info, m, defaultPrefix, caseSens)
+          }
+        }
+      })
+    } else if (isMemberAccess) {
+      // TODO how to avoid allocating a new string? buffer of immutable string slices?
```

Comment:
btw, don't forget the `.substring` memory leak http://stackoverflow.com/questions/15612157 (I think it's fixed Java 7+)